velox/type/tests/TimestampTest.cpp

```diff
@@ -299,37 +299,152 @@ TEST(TimestampTest, outOfRange) {
-  std::tm tm;
-  ASSERT_FALSE(epochToUtc(-(1ll << 60), tm));
-  ASSERT_FALSE(epochToUtc(1ll << 60, tm));
+void checkTm(const std::tm& actual, const std::tm& expected) {
+  ASSERT_EQ(expected.tm_year, actual.tm_year);
+  ASSERT_EQ(expected.tm_yday, actual.tm_yday);
+  ASSERT_EQ(expected.tm_mon, actual.tm_mon);
+  ASSERT_EQ(expected.tm_mday, actual.tm_mday);
+  ASSERT_EQ(expected.tm_wday, actual.tm_wday);
+  ASSERT_EQ(expected.tm_hour, actual.tm_hour);
+  ASSERT_EQ(expected.tm_min, actual.tm_min);
+  ASSERT_EQ(expected.tm_sec, actual.tm_sec);
 }
 
-TEST(TimestampTest, randomEpochToUtc) {
+std::string tmToString(
+    const std::tm& tmValue,
+    const std::string& format,
+    int32_t nanos,
+    const TimestampToStringOptions& options) {
+  auto width = static_cast<int>(options.precision);
+  auto value =
+      options.precision == TimestampToStringOptions::Precision::kMilliseconds
+      ? nanos / 1'000'000
+      : nanos;
+
+  std::ostringstream oss;
+  oss << std::put_time(&tmValue, format.c_str());
+
+  if (options.mode != TimestampToStringOptions::Mode::kDateOnly) {
+    oss << '.' << std::setfill('0') << std::setw(width) << value;
+  }
+
+  return oss.str();
+}
+
+TEST(TimestampUtilTest, epochToUtc) {
+  std::tm tm{};
+  ASSERT_FALSE(Timestamp::epochToUtc(-(1ll << 60), tm));
+  ASSERT_FALSE(Timestamp::epochToUtc(1ll << 60, tm));
+}
+
+TEST(TimestampUtilTest, randomEpochToUtc) {
   uint64_t seed = 42;
   // seed = std::random_device{}();
   std::default_random_engine gen(seed);
   std::uniform_int_distribution<time_t> dist(
       std::numeric_limits<time_t>::min(), std::numeric_limits<time_t>::max());
-  std::tm actual, expected;
+  std::tm actual{}, expected{};
+  for (int i = 0; i < 10000; ++i) {
+    auto epoch = dist(gen);
+    SCOPED_TRACE(fmt::format("epoch={}", epoch));
+    if (gmtime_r(&epoch, &expected)) {
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
+    } else {
+      ASSERT_FALSE(Timestamp::epochToUtc(epoch, actual));
+    }
+  }
+}
+
+TEST(TimestampUtilTest, tmToStringDateOnly) {
+  uint64_t seed = 42;
+  std::default_random_engine gen(seed);
+  std::uniform_int_distribution<time_t> dist(
+      std::numeric_limits<time_t>::min(), std::numeric_limits<time_t>::max());
+  std::tm actual{}, expected{};
+
+  TimestampToStringOptions options;
+  options.mode = TimestampToStringOptions::Mode::kDateOnly;
+  for (int i = 0; i < 10000; ++i) {
+    auto epoch = dist(gen);
+    SCOPED_TRACE(fmt::format("epoch={}", epoch));
+    if (gmtime_r(&epoch, &expected)) {
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
+
+      auto actualStr = Timestamp::tmToString(actual, 0, options);
+      auto expectedStr = tmToString(expected, "%F", 0, options);
+      ASSERT_EQ(expectedStr, actualStr);
+
+    } else {
+      ASSERT_FALSE(Timestamp::epochToUtc(epoch, actual));
+    }
+  }
+}
+
+TEST(TimestampUtilTest, tmToStringTimeOnly) {
```

Comment:
This test case has a lot of overlap with the previous one. Would it make sense to expand the testing section of the previous test case instead, i.e. add something like checkTmString that would generate strings with different precisions and modes and verify they all correct?

Reply:
I've refactored the test case and it looks cleaner now.